### PR TITLE
Gcp haas alert websocket

### PR DESCRIPTION
**Pull request by Michael7371**
This pull request sets up the basis for the Haas alert web socket module. It includes a module to handle rest requests to the active Haas alert API, as well as connectors to the CDOT GCP instance to test connections to pub/sub and secret manager instances. Unit tests are implemented for all supplied code, and instructions on how to run them are supplied in the README. I will separately send both of you the gcp_credentials.json file and the filled in .env file.

### CODE REVIEW

File: haas_websocket/main.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        if data['type'] == 'location':
+            path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('LOCATION_TOPIC'))
+            returnMessage = 'location'
+        if path and returnMessage:
+            publishMessage(publisher,path,encoded)
+            published = True
+            
+    return returnMessage, published
+
+def running():
```
Comment (jacob6838) [inline haas_websocket/main.py:51]: Why is this here? Is this just not fully implemented yet?

File: haas_websocket/main.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    path = None
+    published = False
+    # Filters the messages and publishes messages to specified topic
+    if 'keepAlive' in data:
+        returnMessage = 'keepAlive'
+    if 'type' in data:       
+        if data['type'] == 'point':
+            path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('POINT_TOPIC'))
+            returnMessage = 'point'
+        if data['type'] == 'heartbeat':
```
Comment (jacob6838) [inline haas_websocket/main.py]: should be elif

File: haas_websocket/main.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    encoded = jsonData.encode("utf-8")
+    data = json.loads(jsonData)
+    
+    returnMessage = None
+    path = None
+    published = False
+    # Filters the messages and publishes messages to specified topic
+    if 'keepAlive' in data:
+        returnMessage = 'keepAlive'
+    if 'type' in data:
```
Comment (jacob6838) [inline haas_websocket/main.py]: can eliminate this if statement using 
```
data_type = data.get('type')
```
then data_type will either be data['type'] or None if it does not exist

File: haas_websocket/main.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    returnMessage = None
+    path = None
+    published = False
+    # Filters the messages and publishes messages to specified topic
+    if 'keepAlive' in data:
+        returnMessage = 'keepAlive'
+    if 'type' in data:       
+        if data['type'] == 'point':
+            path = publisher.topic_path(os.getenv('PROJECT_ID'), os.getenv('POINT_TOPIC'))
+            returnMessage = 'point'
```
Comment (jacob6838) [inline haas_websocket/main.py]: you could make a dictionary of return messages, something like
```
RETURN_MESSAGES = {'point': 'point', 'heartbeat': 'heartbeat', ...}
returnMessage = RETURN_MESSAGES.get(data_type)
```

File: haas_websocket/main.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    return True
+
+def startWebsocket(publisher):
+    rest = restSignIn()
+    b_token = rest.getToken()
+    ws_endpoint = os.getenv('HAAS_WSS_ENDPOINT')
+    ws = websocket.create_connection(ws_endpoint+b_token)
+    while running() == True:
+        checkPass = rest.checkPassword()
+        if checkPass == True:
```
Comment (jacob6838) [inline haas_websocket/main.py]: request using guard clauses here. Simply check the negative case, where checkPass == False, and then run the code in the else if it is. Otherwise, leave the if statement and continue. This will reduce unecessary indentation
```
if rest.checkPassword() == False:
    ws.close()
    b_token = rest.tokenUpdate() # updates local token from the secret manager
    ws = websocket.create_connection(ws_endpoint+b_token) # restarts the websocket

... rest of code ...

File: haas_websocket/main.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    return returnMessage, published
+
+def running():
+    return True
+
+def startWebsocket(publisher):
+    rest = restSignIn()
+    b_token = rest.getToken()
+    ws_endpoint = os.getenv('HAAS_WSS_ENDPOINT')
+    ws = websocket.create_connection(ws_endpoint+b_token)
```
Comment (jacob6838) [inline haas_websocket/main.py:58]: create a variable for ws_endpoint+b_token

File: haas_websocket/main.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        checkPass = rest.checkPassword()
+        if checkPass == True:
+            checkToken = rest.checkToken()
+            if checkToken == True: # checks if local bearer_token is the same as the gcp token
+                try:
+                    result = ws.recv() 
+                    print(result)
+                    msg_type, published = filterMessage(result,publisher)
+                    if published == True:
+                        print(f"(INFO) Successfully pushed {msg_type} message pub/sub. \nFull Message: {result}")
```
Comment (jacob6838) [inline haas_websocket/main.py]: we like to use a logging library. Just 
```
import logging
logging.error('error message')
logging.warning('warning message')
logging.info('info message')
```
I would only use print statements if it is a message to the user in a script

File: haas_websocket/main.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+def running():
+    return True
+
+def startWebsocket(publisher):
+    rest = restSignIn()
+    b_token = rest.getToken()
+    ws_endpoint = os.getenv('HAAS_WSS_ENDPOINT')
+    ws = websocket.create_connection(ws_endpoint+b_token)
+    while running() == True:
```
Comment (jacob6838) [inline haas_websocket/main.py:59]: Be careful with infinite loops. Expecially without an exit, even when there are errors. You should exit after a few failures

File: README.md [hunk 9]
```diff
@@ -1,2 +1,41 @@
-# HAASAlert
-CDOT solution to ingest HAASAlert data for emergency services data within GCP
+## Project Description
```
Comment (jacob6838) [inline README.md:1]: Can we get a brief diagram of what the system looks like? where the websocket is hosted and the refresh token system?

File: haas_websocket/haas_password_generator.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+
+
+alphabets = list(string.ascii_letters)
+digits = list(string.digits)
+special_characters = list("!@#$%^&*()")
+characters = list(string.ascii_letters + string.digits + "!@#$%^&*()")
+
+
+def gen_password(length):
```
Comment (jacob6838) [inline haas_websocket/haas_password_generator.py]: Can replace this entire function with one line:
```
''.join(random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(N))
```
https://stackoverflow.com/questions/2257441/random-string-generation-with-upper-case-letters-and-digits

File: haas_websocket/haas_rest_handler.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            client = secretmanager.SecretManagerServiceClient()
+            parent = client.secret_path(self.id, secret_id)
+            sv_list = client.list_secret_versions
+            parent_list = sv_list(request={"parent": parent})
+            for version in parent_list:
+                if (version.state == 1):
+                    response = client.access_secret_version(request={"name": version.name})
+                    decoded = response.payload.data.decode("UTF-8")
+                    return decoded
+        except Exception:
```
Comment (jacob6838) [inline haas_websocket/haas_rest_handler.py]: log error message. Usually better to catch specific exceptions

Comment (jacob6838) [reply haas_websocket/haas_rest_handler.py]: some exceptions, like a lack of internet, would be a cause to exit the entire program

File: haas_websocket/haas_rest_handler.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            return None
+        try:
+            client = secretmanager.SecretManagerServiceClient()
+            encoded_payload = payload.encode("UTF-8")
+            parent = f"projects/{self.id}/secrets/{secret_id}"
+            response = client.add_secret_version(
+                request={"parent": parent, "payload": {"data": encoded_payload}}
+            )
+            return response
+        except Exception:
```
Comment (jacob6838) [inline haas_websocket/haas_rest_handler.py:48]:
```
except Exception as e:
```
then log the error message with 
```
str(e)
```

File: haas_websocket/haas_rest_handler.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        self.api_username = self.secretManagerGet(self.username_key)
+        self.api_password = self.secretManagerGet(self.password_key)
+
+        r = requests.post(self.api_endpoint + "oauth/token",
+                    json={"grant_type": "password",
+                    "username":self.api_username,
+                    "password":self.api_password})
+        json_response = json.loads(r.content)
+
+        self.b_token = json_response["access_token"]
```
Comment (jacob6838) [inline haas_websocket/haas_rest_handler.py]: use .get to access values safely

File: haas_websocket/haas_rest_handler.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+    def signIn(self):
+        self.api_username = self.secretManagerGet(self.username_key)
+        self.api_password = self.secretManagerGet(self.password_key)
+
+        r = requests.post(self.api_endpoint + "oauth/token",
+                    json={"grant_type": "password",
+                    "username":self.api_username,
+                    "password":self.api_password})
+        json_response = json.loads(r.content)
```
Comment (jacob6838) [inline haas_websocket/haas_rest_handler.py]: if this request fails, shouldn't we return something? Or can this never fail?

File: haas_websocket/haas_rest_handler.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        return self.b_token
+
+
+    def signOut(self):
+        r = requests.post(self.api_endpoint+'oauth/revoke')
+
+        if r.status_code == 200:
+            return True
+        else:
+            print("(ERROR) FAILED TO CONNECT TO HAAS ALERT AND SIGN OUT")
```
Comment (jacob6838) [inline haas_websocket/haas_rest_handler.py]: use response to improve error message

File: haas_websocket/main.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
+import os
+import sys
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
Comment (jacob6838) [inline haas_websocket/main.py]: what does this do?

File: point.json [hunk 17]
```diff
@@ -0,0 +1 @@
+{"accuracy":0,"alt":2279.5171,"course":280,"group_id":"3690","id":"05da255d-9079-4a9f-af78-ac7463aab412","is_active":false,"lat":39.705635,"lon":-105.27304367,"speed":30.3708,"thing":{"external_id":"410037000c47373336373936","id":"ead04cde-fc91-42c1-af6e-2a6d7440292a"},"timestamp":"2022-02-23T21:19:24.00Z","type":"point"}
```
Comment (jacob6838) [inline point.json:1]: "thing"??

Comment (Michael7371) [reply point.json:1]: yeah they use thing to denote whatever the object that is sending this message is. It could be an engine, station, etc.

File: haas_websocket/main.py (hunk 16 above)
Comment (Michael7371) [reply haas_websocket/main.py]: Not needed anymore, used to be useful for imports but now since it's project based it's not needed. Will remove.

File: point.json (hunk 17 above)
Comment (jacob6838) [reply point.json:1]: can this be enumerated or slightly less generic? like data?

Comment (jacob6838) [reply point.json:1]: oh wait is this from HaasAlert? lame

File: haas_websocket/haas_rest_handler.py (hunk 15 above)
Comment (Michael7371) [reply haas_websocket/haas_rest_handler.py]: okay, is adding r.reason good?

File: haas_websocket/haas_rest_handler.py (hunk 14 above)
Comment (Michael7371) [reply haas_websocket/haas_rest_handler.py]: Yeah I should add some error catching there.

File: haas_websocket/haas_rest_handler.py (hunk 15 above)
Comment (jacob6838) [reply haas_websocket/haas_rest_handler.py]: sure, whatever the response message is

File: haas_websocket/haas_rest_handler.py (hunk 13 above)
Comment (Michael7371) [reply haas_websocket/haas_rest_handler.py]: like this: json_response.get("access_token") ?

File: haas_websocket/haas_rest_handler.py (hunk 12 above)
Comment (Michael7371) [reply haas_websocket/haas_rest_handler.py:48]: Added

File: haas_websocket/haas_rest_handler.py (hunk 13 above)
Comment (jacob6838) [reply haas_websocket/haas_rest_handler.py]: yeah, then the return will be None if the key doesn't exist. Only use when you have other error handling that will handle None though

File: haas_websocket/main.py (hunk 6 above)
Comment (Michael7371) [reply haas_websocket/main.py:58]: The reason I do this is just so when b_token updates I wouldn't have to also update the new variable. Is there a way to make a new variable that references b_token so when b_token updates the new variable would update with it?

Comment (jacob6838) [reply haas_websocket/main.py:58]: makes sense, no need to change it then

File: haas_websocket/main.py (hunk 1 above)
Comment (Michael7371) [reply haas_websocket/main.py:51]: The running method? It is there for unit testing, couldn't figure out a better way to exit the while loop.

File: haas_websocket/main.py (hunk 4 above)
Comment (Michael7371) [reply haas_websocket/main.py]: Yeah, I guess how would I change the topic for each message with a dict?

Comment (Michael7371) [reply haas_websocket/main.py]: for the path

File: README.md (hunk 9 above)
Comment (Michael7371) [reply README.md:1]: Yeah, I can work on that

File: haas_websocket/main.py (hunk 16 above)
Comment (Michael7371) [reply haas_websocket/main.py]: Actually, turns out this is useful. It can't perform imports without setting the working path to the path of the repo.

File: haas_websocket/haas_password_generator.py (hunk 10 above)
Comment (Michael7371) [reply haas_websocket/haas_password_generator.py]: Yeah, I kind of want to keep it for now and just re-work it later with the other work item. I do need one of each of those types.

File: haas_websocket/main.py (hunk 4 above)
Comment (jacob6838) [reply haas_websocket/main.py]: another dictionary, this time with values being the keys for the environment variables?

File: haas_websocket/main.py [hunk 18]
```diff
@@ -57,33 +57,33 @@ def startWebsocket(publisher):
...
-                ws = websocket.create_connection(ws_endpoint+b_token) # restarts the websocket
-        else:
+        check_pass = rest.checkPassword()
+        check_token = rest.checkToken()
+        if check_pass == False:
             ws.close()
             rest.passwordUpdate() # updates local token from the secret manager
             b_token = rest.signIn()
             ws = websocket.create_connection(ws_endpoint+b_token) # restarts the websocket
+        if check_token == False:
```
Comment (jacob6838) [inline haas_websocket/main.py:67]: should these be else if statements? Or is it ok if it goes into the first two in a row?

Comment (Michael7371) [reply haas_websocket/main.py:67]: I think they should both be checked.

**Comment by jacob6838** [comment]
I think the only remaining task is the system diagram

**Comment by Michael7371** [comment]
> 

Yeah, I forgot about that. Will start on it now.
